### ISSUE TITLE
Make executor variable in slick.Async.Executor's 2nd apply method pluggable

### ISSUE THREAD

**Issue by jayeve** [slick/slick#2046]
**DESCRIPTION:**

The default `AsyncExecutor` on offer in slick requires you use a `java.util.concurrent.ThreadPoolExecutor` as the [underlying executor](https://github.com/slick/slick/blob/v3.3.2/slick/src/main/scala/slick/util/AsyncExecutor.scala#L110) for running db actions. This could be programmable.

It would be nice to have the intelligent queuing logic in the default `AsyncExecutor` implementation as well as on ability to override the `run` method in the thread executor. Specifically, I need a way to propagate context from thread to thread, similar to how this is done in [Finagle FuturePool](https://github.com/twitter/util/blob/54213cb/util-core/src/main/scala/com/twitter/util/FuturePool.scala#L128-L150). I've written a custom [TwitterLocalThreadPoolExecutor](https://github.com/jayeve/slick-executor/blob/master/TwitterLocalThreadPoolExecutor.scala) with a `run` method doing just that. Currently, I'm using the scala reflection library to swap out the executor within the `slick.util.AsyncExecutor` with my custom one.

I think the easier solution would be to add an apply method to the `AsyncExecutor` object with a `execute` function as an optional parameter.

https://github.com/slick/slick/blob/93ccca0881a831389ecb7aff2892b8a482519eef/slick/src/main/scala/slick/util/AsyncExecutor.scala#L67

```scala
def apply(
  name: String,
  minThreads: Int,
  maxThreads: Int,
  queueSize: Int,
  maxConnections: Int = Integer.MAX_VALUE,
  keepAliveTime: Duration = 1.minute,
  registerMbeans: Boolean = false,
  customExecutor: Option[Runnable => Unit] 
): AsyncExecutor = {
  ...
}
```

Then, override the `execute` method on the [anonymous instance](https://github.com/slick/slick/blob/v3.3.2/slick/src/main/scala/slick/util/AsyncExecutor.scala#L110) of the `executor`

```scala
override def execute(runnable: Runnable): Unit = { 
  customExecutor match { 
    case Some(executeFn) => executeFn(runnable)
    case None => super.execute(runnable)
  } 
}  
```

Thoughts on this approach?

**Comment by sluicing** [comment]
I'm interested in this issue as well.

> Thoughts on this approach?

This approach seems workable to me, though I also looked at approaches like that used in springs TaskDecorator, which provide an opportunity to wrap the created Runnable with custom logic, similar to https://moelholm.com/blog/2017/07/24/spring-43-using-a-taskdecorator-to-copy-mdc-data-to-async-threads

I don't have a strong preference to either approach myself, and I might not understand the issues well enough to know if the decorator approach will allow everything that @jayeve wants to do.

> Currently, I'm using the scala reflection library to swap out the executor within the slick.util.AsyncExecutor 
> with my custom one.

Care to share what this looks like?

**Comment by hvesalai** [comment]
How would you do the queuing?

There are requirements for the number of threads (see code starting at https://github.com/slick/slick/blob/v3.3.2/slick/src/main/scala/slick/util/AsyncExecutor.scala#L94)

**Comment by hvesalai** [comment]
And have you analyzed how this would relate to the tasks that already have a connection. They must be run before any other tasks (because the other tasks might not be able to get a connection unless the task holding it releases it) 
https://github.com/slick/slick/blob/v3.3.2/slick/src/main/scala/slick/util/AsyncExecutor.scala#L200

**Comment by sluicing** [comment]
> How would you do the queuing?
> And have you analyzed how this would relate to the tasks that already have a connection.

You're right that replacing the executor isn't workable without reworking queuing and priority, etc.

I'm only interested in setting up thread local items for log4j MDC, to propagate MDC information into the runnable used to execute slick items.  After the MDC information is set up, I would delegate to the original executor for the rest of the work.

The original poster was also interested mostly in setting up thread local variables, it seems; they too might be delegating to the original execution context after those variables are set up.

**Comment by jayeve** [comment]
> Care to share what this looks like?

@sluicing I put a mock-up in this [gist](https://gist.github.com/jayeve/252aa0939538b0507e829cf8841b7866). The `DBConfig` object is a container for the slick `AsyncExecutor`.

Our purpose is to transfer `ThreadLocal` state between twitter Futures and the Slick threads that execute SQL queries. This enables us to connect server requests to underlying SQL queries in our distributed tracing. It's been working for us since mid 2019.

**Comment by sluicing** [comment]
Thanks!  I'll take a look.

We have (I think) almost exactly the same use case, trying to connect a thread identifier from a twitter future service a request into the slick thread that executes queries.
